File: problem-084/solution.go
package problem084

// countIslands counts number of islands in given grid
// of land (represented in 1) and water (represented by 0).
// Island is a group of 1 that are surrounded by 0.
func countIslands(grid [][]int) int {
	if len(grid) == 0 {
		return 0
	}

	count := 0
	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[0]); j++ {
			if grid[i][j] != 1 {
				continue
			}
			dfs(grid, i, j)
			count += 1
		}
	}

	return count
}

// dfs visits neighboring cells recursively until perimeter
// or already visited cell is found.
func dfs(grid [][]int, row, col int) {
	// check if out of bound and cell is an unvisited islan
	if row < 0 || col < 0 || row >= len(grid) || col >= len(grid[0]) || grid[row][col] != 1 {
		return
	}

	// mark cell as visited
	grid[row][col] = 2

	// check neighbors
	dfs(grid, row+1, col)
	dfs(grid, row-1, col)
	dfs(grid, row, col+1)
	dfs(grid, row, col-1)
}

File: problem-084/solution.py
from typing import List


def count_islands(grid: List[List[int]]) -> int:
    """
    1. Iterate each cell.
    2. If cell is 1:
        1. Iterate each neighbor with dfs until water found or already visited cell found.
    """
    if not grid:
        return 0

    count = 0
    for row in range(len(grid)):
        for col in range(len(grid[0])):
            if grid[row][col] == 1:
                dfs(grid, row, col)
                count += 1
    return count


def dfs(grid: List[List[int]], row: int, col: int) -> None:
    if row < 0 or col < 0 or row >= len(grid) or col >= len(grid[0]) or grid[row][col] != 1:
        return

    grid[row][col] = 2

    dfs(grid, row+1, col)
    dfs(grid, row-1, col)
    dfs(grid, row, col+1)
    dfs(grid, row, col-1)


test_grid: List[List[int]] = [
    [1, 0, 0, 0, 0],
    [0, 0, 1, 1, 0],
    [0, 1, 1, 0, 0],
    [0, 0, 0, 0, 0],
    [1, 1, 0, 0, 1],
    [1, 1, 0, 0, 1]
]
expected_result = 4
assert count_islands(test_grid) == expected_result